lmdeploy/serve/openai/api_server.py

```diff
@@ -973,6 +974,18 @@ async def startup_event():
         print(f'Service registration failed: {e}')
 
 
+class ConcurrencyLimitMiddleware(BaseHTTPMiddleware):
+
+    def __init__(self, app: FastAPI, max_concurrent_requests: int):
+        super().__init__(app)
+        self.semaphore = asyncio.Semaphore(max_concurrent_requests)
+
+    async def dispatch(self, request: Request, call_next):
```

Comment:
How does it work? Any reference?

Reply:
It is used in `BaseHTTPMiddleware.__call__`